backend/ethereum/channel/deploy_test.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package channel_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	ethchannel "perun.network/go-perun/backend/ethereum/channel"
+	"perun.network/go-perun/backend/ethereum/channel/test"
+	pkgtest "perun.network/go-perun/pkg/test"
+)
+
+func TestDeployReorgResistance(t *testing.T) {
+	rng := pkgtest.Prng(t)
+
+	txFinalityDepth := uint64(TxFinalityDepthMin + rng.Intn(TxFinalityDepthMax-TxFinalityDepthMin+1))
+	t.Logf("txFinalityDepth = %v", txFinalityDepth)
+	s := test.NewSimSetup(t, rng, txFinalityDepth, 0)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	deploy := make(chan error)
+	go func() {
+		_, err := ethchannel.DeployAdjudicator(ctx, *s.CB, s.TxSender.Account)
+		deploy <- err
+	}()
+
+	commit := make(chan error)
+	go func() {
+		time.Sleep(100 * time.Millisecond) // Ensure that commits are applied after deployment transaction.
+		for i := 0; i < int(txFinalityDepth); i++ {
+			select {
+			case <-deploy:
+				commit <- errors.New("deploy done before commit")
+				return
+			default:
+			}
+			s.SimBackend.Commit()
+		}
+		commit <- nil
+	}()
+
+	select {
```

Comment:
This waits for *either* deploy or commit, so it never checks if it actually deployed *and* committed.  
I think the second routine is a bit overkill. You can use this if you want:  
```go
for i := 0; i < int(txFinalityDepth-1); i++ {
	time.Sleep(100 * time.Millisecond) // Give an erroneous deploy function time to finish.
	select {
	case <-deploy:
		t.Error("deployment returned too early")
		t.FailNow()
	default:
	}
	s.SimBackend.Commit()
}
assert.NoError(t, <-deploy)
```
Note: The `-1` should be here as well since `deployContract` already seems to mint one block.  
Putting the sleep into the loop would ensure that the `Commit()` calls have time to be forwarded to the sub before checking for its return.

Reply:
Thanks for catching this mistake. I refactored several times and apparently didn't do it properly.

I am now following your approach with one exception. Manually reducing the `txFinalityDepth` by 1 to account for the fact that the deploy transaction usually automatically induces one block being mined isn't very clean. Instead, I modified the simBackend to completely allow for manual control. Hence, we can now control that the deployment does not automatically produce a new block and therefore we can use `txFinalityDepth` as we should.

Reply:
Yes good idea.